rasa/server.py

```diff
@@ -476,6 +479,124 @@ async def hello(request: Request):
+        # query parameter.
+        if not callback_url:
+            return await f(request, *args, **kwargs)
+
+        async def wrapped() -> None:
+            try:
+                result: HTTPResponse = await f(request, *args, **kwargs)
+                payload = dict(
+                    data=result.body, headers={"Content-Type": result.content_type}
+                )
+                logger.debug(
+                    "Asynchronous processing of request was successful. "
+                    "Sending result to callback URL."
+                )
+
+            except Exception as e:
+                if not isinstance(e, ErrorResponse):
+                    logger.error(e)
+                    e = ErrorResponse(
+                        HTTPStatus.INTERNAL_SERVER_ERROR,
+                        "UnexpectedError",
+                        f"An unexpected error occurred. Error: {e}",
+                    )
+                # If an error happens, we sent the error payload to the `callback_url`
+                payload = dict(json=e.error_info)
+                logger.debug(
+                    "Error happened when processing request asynchronously. "
+                    "Sending error to callback URL."
+                )
+            async with aiohttp.ClientSession() as session:
+                await session.post(callback_url, raise_for_status=True, **payload)
+
+        # Run the request in the background on the event loop
+        request.app.add_task(wrapped())
+
+        # The incoming request will return immediately with a 204
+        return response.empty()
+
+    return decorated_function
+
+
+def heavy_computation(f: Callable[..., Coroutine]) -> Callable:
+    """Decorator which runs request on a separate thread.
+
+    Some requests (e.g. training or cross-validation) are computional intense requests.
+    This means that they will block the event loop and hence the processing of other
+    requests. This decorator can be used to process these requests on a separate thread
+    to avoid blocking the processing of incoming requests.
+
+    Args:
+        f: The request handler function which should be decorated.
+
+    Returns:
+        The decorated function.
+    """
+
+    @wraps(f)
+    async def decorated_function(
+        request: Request, *args: Any, **kwargs: Any
+    ) -> HTTPResponse:
+        # Use a sync wrapper for our `async` function as `run_in_executor` only supports
+        # sync functions
+        def run() -> HTTPResponse:
+            # This is a new thread, so we need to create and set a new event loop
+            thread_loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(thread_loop)
+
+            try:
+                return thread_loop.run_until_complete(f(request, *args, **kwargs))
+            finally:
+                thread_loop.close()
+
+        # Run decorated function in thread (`None` will use the `ThreadPoolExecutor`)
```

Comment:
Better to explicitly set it than rely on a comment I think.

Reply:
we rely on the Python default in that case 😄  Will make it explicit